odl/tomo/backends/astra_cuda.py

```diff
@@ -0,0 +1,199 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Backen for ASTRA using CUDA"""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+
+# External
+import numpy as np
+
+try:
+    import astra
+    if astra.astra.use_cuda():
+        ASTRA_CUDA_AVAILABLE = True
+except ImportError:
+    ASTRA_CUDA_AVAILABLE = False
+
+# Internal
+from odl.discr.lp_discr import DiscreteLp, DiscreteLpVector
+from odl.tomo.backends.astra_setup import (astra_projection_geometry,
+                                           astra_volume_geometry,
+                                           astra_projector, astra_data,
+                                           astra_algorithm, astra_cleanup)
+from odl.tomo.geometry.geometry import Geometry
+
+
+__all__ = ('astra_gpu_forward_projector_call',
+           'astra_gpu_backward_projector_call',
+           'ASTRA_CUDA_AVAILABLE')
+
+
+# TODO: rename gpu to cuda?
+def astra_gpu_forward_projector_call(vol_data, geometry, proj_space, out=None):
+    """Run an ASTRA forward projection on the given data using the GPU.
+
+    Parameters
+    ----------
+    vol_data : `DiscreteLpVector`
+        Volume data to which the projector is applied
+    geometry : `Geometry`
+        Geometry defining the tomographic setup
+    proj_space : `DiscreteLp`
+        Space to which the calling operator maps
+    out : `DiscreteLpVector` or `None`, optional
+        Vector in the projection space to which the result is written
+        Default: `None`
+
+    Returns
+    -------
+    out : ``proj_space`` element
+        Projection data resulting from the application of the projector
+    """
+    if not isinstance(vol_data, DiscreteLpVector):
+        raise TypeError('volume data {!r} is not a `DiscreteLp.Vector` '
+                        'instance.'.format(vol_data))
+    if not isinstance(geometry, Geometry):
+        raise TypeError('geometry  {!r} is not a `Geometry` instance.'
+                        ''.format(geometry))
+    if not isinstance(proj_space, DiscreteLp):
+        raise TypeError('projection space {!r} is not a `DiscreteLp` '
+                        'instance.'.format(proj_space))
+
+    if vol_data.space.grid.ndim != geometry.ndim:
+        raise ValueError('dimensions {} of volume data and {} of geometry '
+                         'do not match.'
+                         ''.format(vol_data.space.grid.ndim, geometry.ndim))
+
+    ndim = vol_data.space.grid.ndim
+    # Create astra geometries
+    vol_geom = astra_volume_geometry(vol_data.space)
+    proj_geom = astra_projection_geometry(geometry)
+
+    # Create ASTRA data structures
+
+    # In the case dim == 3, we need to swap axes, so can't perform the FP
```

Comment:
The logic is somewhat complicated here (Non-local, specifically). Any chance to simplify it?